File: example-android-project/app/src/main/java/io/github/gofaith/faithdroid/FViews/FViewPager.java
package io.github.gofaith.faithdroid.FViews;

import android.os.Bundle;
import android.support.design.widget.TabLayout;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentStatePagerAdapter;
import android.support.v4.view.ViewCompat;
import android.support.v4.view.ViewPager;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import org.json.JSONArray;
import org.json.JSONObject;
import org.json.JSONTokener;

import java.util.ArrayList;
import java.util.List;

import faithdroid.Faithdroid;
import io.github.gofaith.faithdroid.UI.AttrGettable;
import io.github.gofaith.faithdroid.UI.AttrSettable;
import io.github.gofaith.faithdroid.UI.DemoObjectFragment;
import io.github.gofaith.faithdroid.UI.UIController;

public class FViewPager extends FView implements AttrSettable, AttrGettable {
    private FaithCollectionPagerAdapter adapter;
    public ViewPager v;
    public List<FPage> pages = new ArrayList<>();
    public String onCreateView,onGetCount,onPageSelected;
    public FTabLayout bindFTabLayout;
    public FViewPager(UIController controller) {
        parentController = controller;
        v = new ViewPager(parentController.activity);
        v.addOnPageChangeListener(new ViewPager.OnPageChangeListener() {
            @Override
            public void onPageScrolled(int i, float v, int i1) {

            }

            @Override
            public void onPageSelected(int i) {
                if (onPageSelected != null) {
                    Faithdroid.triggerEventHandler(onPageSelected, String.valueOf(i));
                }
            }

            @Override
            public void onPageScrollStateChanged(int i) {

            }
        });
        view=v;
        parseSize("[-1,-1]");
    }

    @Override
    public String getAttr(String attr) {
        String str = getUniversalAttr(attr);
        if (str != null) {
            return str;
        }
        switch (attr){
            // --------------------------------------------------
        }
        return "";
    }

    @Override
    public void setAttr(String attr, final String value) {
        if (setUniversalAttr(attr, value)) {
            return;
        }
        switch (attr){
            // ----------------------------------------------------------------------------
            case "Pages":
                Log.d(TAG, "setAttr: "+value);
                try {
                    JSONArray array = (JSONArray) (new JSONTokener(value).nextValue());
                    for(int i=0;i<array.length();i++){
                        JSONObject object = array.getJSONObject(i);
                        FPage fPage=new FPage();
                        fPage.vid = object.getString("VID");
                        pages.add(fPage);
                    }
                    if(adapter==null) {
                        adapter = new FaithCollectionPagerAdapter(parentController.activity.getSupportFragmentManager(), this);
                        v.setAdapter(adapter);
                    }
                } catch (Exception e) {
                    e.printStackTrace();
                }
                break;
            case "TabLayout":
                FView o = parentController.viewmap.get(value);
                if (o==null)break;
                FTabLayout fTabLayout=(FTabLayout)o;
                bindFTabLayout=fTabLayout;
                fTabLayout.v.setupWithViewPager(v);
                break;
            case "OnCreateView":
                onCreateView = value;
                if (adapter==null&&onGetCount!=null) {
                    adapter = new FaithCollectionPagerAdapter(parentController.activity.getSupportFragmentManager(), this);
                    v.setAdapter(adapter);
                }
                break;
            case "OnGetCount":
                onGetCount=value;
                if (adapter==null&&onCreateView!=null) {
                    adapter = new FaithCollectionPagerAdapter(parentController.activity.getSupportFragmentManager(), this);
                    v.setAdapter(adapter);
                }
                break;
            case "CurrentItem":
                try {
                    JSONArray array = (JSONArray) (new JSONTokener(value).nextValue());
                    int index = array.getInt(0);
                    int is = array.getInt(1);
                    v.setCurrentItem(index,is==1);
                } catch (Exception e) {
                    e.printStackTrace();
                }
                break;
            case "OnPageSelected":
                onPageSelected = value;
                break;
        }
    }
    // ------------------
   public class FPage{
        public String vid;
    }
    class FaithCollectionPagerAdapter extends FragmentStatePagerAdapter {
        private final FViewPager fviewPager;

        public FaithCollectionPagerAdapter(FragmentManager fm, FViewPager fViewPager) {
            super(fm);
            this.fviewPager =fViewPager;
        }

        @Override
        public Fragment getItem(int i) {
            Fragment fragment = new DemoObjectFragment().setFviewPager(fviewPager);
            Bundle args = new Bundle();
            // Our object is just an integer :-P
            args.putInt(DemoObjectFragment.ARG_OBJECT, i);
            fragment.setArguments(args);
            return fragment;
        }
        @Override
        public int getCount() {
            if (fviewPager.pages.size()==0&&onGetCount!=null&&onCreateView!=null) {
                try {
                    return Integer.parseInt(Faithdroid.triggerEventHandler(onGetCount, ""));
                } catch (Exception e) {
                }
            }
            return fviewPager.pages.size();
        }
        @Override
        public CharSequence getPageTitle(int position) {
            if (bindFTabLayout==null||bindFTabLayout.tabsList.size()<=position){
                return "Page "+position;
            }
            return bindFTabLayout.tabsList.get(position);
        }
    }
}

File: android/app/src/main/java/io/github/stevenzack/faithdroidapp/FViews/FViewPager.kt
package io.github.stevenzack.faithdroidapp.FViews

import android.content.ContentValues.TAG
import android.os.Bundle
import android.util.Log
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentManager
import androidx.fragment.app.FragmentStatePagerAdapter
import androidx.viewpager.widget.ViewPager
import faithdroid.Faithdroid
import io.github.stevenzack.faithdroidapp.UI.*
import org.json.JSONArray
import org.json.JSONTokener


class FViewPager(controller: UIController) : FView(), AttrSettable, AttrGettable {
    private var adapter: FaithCollectionPagerAdapter? = null
    var v: ViewPager
    var pages: MutableList<FPage> = ArrayList()
    var onCreateView: String? = null
    var onGetCount: String? = null
    var onPageSelected: String? = null
    var bindFTabLayout: FTabLayout? = null

    init {
        parentController = controller
        v = ViewPager(parentController!!.activity)
        v.addOnPageChangeListener(object : ViewPager.OnPageChangeListener {
            override fun onPageScrolled(i: Int, v: Float, i1: Int) {

            }

            override fun onPageSelected(i: Int) {
                if (onPageSelected != null) {
                    Faithdroid.triggerEventHandler(onPageSelected, i.toString())
                }
            }

            override fun onPageScrollStateChanged(i: Int) {

            }
        })
        view = v
        parseSize("[-1,-1]")
    }

   override fun getAttr(attr: String): String {
        val str = getUniversalAttr(attr)
        if (str != null) {
            return str
        }
        when (attr) {

        }// --------------------------------------------------
        return ""
    }

   override fun setAttr(attr: String, value: String) {
        if (setUniversalAttr(attr, value)) {
            return
        }
        when (attr) {
            // ----------------------------------------------------------------------------
            "Pages" -> {
                Log.d(TAG, "setAttr: $value")
                try {
                    val array = JSONTokener(value).nextValue() as JSONArray
                    for (i in 0 until array.length()) {
                        val `object` = array.getJSONObject(i)
                        val fPage = FPage()
                        fPage.vid = `object`.getString("VID")
                        pages.add(fPage)
                    }
                    if (adapter == null) {
                        adapter =
                            FaithCollectionPagerAdapter(parentController!!.activity.getSupportFragmentManager(), this)
                        v.setAdapter(adapter)
                    }
                } catch (e: Exception) {
                    e.printStackTrace()
                }

            }
            "TabLayout" -> {
                val o = parentController!!.viewmap.get(value) ?: return
                val fTabLayout = o as FTabLayout
                bindFTabLayout = fTabLayout
                fTabLayout.v.setupWithViewPager(v)
            }
            "OnCreateView" -> {
                onCreateView = value
                if (adapter == null && onGetCount != null) {
                    adapter = FaithCollectionPagerAdapter(parentController!!.activity.getSupportFragmentManager(), this)
                    v.setAdapter(adapter)
                }
            }
            "OnGetCount" -> {
                onGetCount = value
                if (adapter == null && onCreateView != null) {
                    adapter = FaithCollectionPagerAdapter(parentController!!.activity.getSupportFragmentManager(), this)
                    v.setAdapter(adapter)
                }
            }
            "CurrentItem" -> try {
                val array = JSONTokener(value).nextValue() as JSONArray
                val index = array.getInt(0)
                val `is` = array.getInt(1)
                v.setCurrentItem(index, `is` == 1)
            } catch (e: Exception) {
                e.printStackTrace()
            }

            "OnPageSelected" -> onPageSelected = value
        }
    }

    // ------------------
    inner class FPage {
        var vid: String? = null
    }

    internal inner class FaithCollectionPagerAdapter(fm: FragmentManager, private val fviewPager: FViewPager) :
        FragmentStatePagerAdapter(fm) {
        override fun getCount():Int{
                if (fviewPager.pages.size == 0 && onGetCount != null && onCreateView != null) {
                    try {
                        return Integer.parseInt(Faithdroid.triggerEventHandler(onGetCount, ""))
                    } catch (e: Exception) {
                    }

                }
                return fviewPager.pages.size
            }

        override fun getItem(i: Int): Fragment {
            val fragment = DemoObjectFragment().setFviewPager(fviewPager)
            val args = Bundle()
            // Our object is just an integer :-P
            args.putInt(DemoObjectFragment.ARG_OBJECT, i)
            fragment.setArguments(args)
            return fragment
        }

        override fun getPageTitle(position: Int): CharSequence {
            return if (bindFTabLayout == null || bindFTabLayout!!.tabsList.size <= position) {
                "Page $position"
            } else bindFTabLayout!!.tabsList[position]
        }
    }
}
